JASP-Engine/JASP/R/binomialtest.R

```diff
@@ -15,224 +15,315 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
 
-BinomialTest <- function(dataset, options, perform, state, ...) {
+BinomialTest <- function(jaspResults, dataset, options, ...) {
 
-	variables <- unlist(options$variables)
+  # Set title
+  jaspResults$title <- "Binomial Test"
 
-	binomResults <- state$binomResults
-  if (is.null(binomResults))
-		binomResults <- .binomialTest(dataset, options, perform, variables)
+  # Read dataset
+  dataset <- .binomReadData(dataset, options)
 
-	descriptPlots <- state$descriptPlots
-  if (options[["descriptivesPlots"]] && is.null(descriptPlots))
-		descriptPlots <- .binomialDescriptivesPlot(dataset, options, perform, variables)
+  # Error checking
+  .binomCheckErrors(dataset, options)
 
-	results <- list()
-	results[["binomial"]] <- .BinomialTable(binomResults, options)
+  # Compute the results
+  binomResults <- .binomComputeResults(jaspResults, dataset, options)
 
-	if (! is.null(descriptPlots)) {
-		results[["descriptives"]] <- descriptPlots
-		results[["descriptives"]][["title"]] <- "Descriptive Plots"
-	}
+  # Output tables and plots
+  .binomTableMain(jaspResults, dataset, options, binomResults)
+  .binomContainerPlots(jaspResults, dataset, options, binomResults)
+  .binomPlotsDescriptive(jaspResults, dataset, options, binomResults)
 
-  if (perform == "init")
-		return(list(results=results, status="inited", state=state))
+  return()
 
-	state <- list(
-		options = options,
-		binomResults = binomResults,
-		descriptPlots = descriptPlots
-	)
+}
 
-  return(list(results=results, status="complete", state=state))
+# Preprocessing functions ----
+.binomReadData <- function(dataset, options) {
+  if (!is.null(dataset)) {
+    return(dataset)
+  } else {
+    return(.readDataSetToEnd(columns.as.factor = options$variables))
+  }
 }
 
-.binomialTest <- function(dataset, options, perform, variables) {
+.binomCheckErrors <- function(dataset, options) {
+
+  # Check if results can be computed
+  if (length(options$variables) == 0) .quitAnalysis("Input at least one variable")
```

Comment:
This causes the default behavior to return an error if the user does not load any variables. We could make this the default, all analyses start with an error message, but perhaps we should discuss this with EJ. I don't think right now this currently very user friendly. The least we could do is create a separate function for cases like this such that the user experience looks better (i.e., a jasp-green box telling them the analysis won't run until they enter a), b) c).

Reply:
yes, good point, I thought I had already changed this. I wanted to change it back to displaying an empty main table. I'll have to look how I can do this while maintaining as much separation as possible between checking for errors and computing results.

Reply:
Maybe just put a jaspHtml with a message stating that one should add some variables? And then return from the (sub)analysis?

Reply:
I have solved it similarly to how it's done now: an empty main table.